lib/wallet/coinselector.js

```diff
@@ -0,0 +1,883 @@
+    }
+
+    // How much money is left after filling outputs.
+    this.change = this.tx.getInputValue() - this.total();
+
+    return this;
+  }
+
+  /**
+   * Initialize selection based on size estimate.
+   */
+
+  async selectEstimate() {
+    // Set minimum fee and do
+    // an initial round of funding.
+    this.fee = CoinSelector.MIN_FEE;
+    this.fund();
+
+    // Add dummy output for change.
+    const change = new Output();
+
+    if (this.changeAddress) {
+      change.script.fromAddress(this.changeAddress);
+    } else {
+      // In case we don't have a change address,
+      // we use a fake p2wpkh output to gauge size.
+      change.script.fromProgram(0, Buffer.allocUnsafe(20));
+    }
+
+    this.tx.outputs.push(change);
+
+    // Keep recalculating the fee and funding
+    // until we reach some sort of equilibrium.
+    do {
+      const size = await this.tx.estimateSize(this.getAccount);
+
+      this.fee = this.getFee(size);
+
+      if (this.maxFee > 0 && this.fee > this.maxFee)
+        throw new FundingError('Fee is too high.');
+
+      // Failed to get enough funds, add more coins.
+      if (!this.isFull())
+        this.fund();
+    } while (!this.isFull() && this.index < this.coins.length);
+  }
+
+  /**
+   * Initiate selection based on a hard fee.
+   */
+
+  selectHard() {
+    this.fee = Math.min(this.hardFee, CoinSelector.MAX_FEE);
+    this.fund();
+  }
+
+  /**
+   * Initiate multi algorithm coin selection.
+   */
+
+  async attemptSelection() {
+    // calculate cost of creating and spending a change output
+    // add dummy output for change.
+    const change = new Output();
+
+    if (this.changeAddress) {
+      change.script.fromAddress(this.changeAddress);
+    } else {
+      // In case we don't have a change address,
+      // we use a fake p2wpkh output to gauge size.
+      change.script.fromProgram(0, Buffer.allocUnsafe(20));
+    }
+
+    const changeOutputSize = change.getSize();
+    const changeOutputCoin = new Coin();
+    changeOutputCoin.script = change.script;
+    const changeSpendingSize =
+      await changeOutputCoin.estimateSpendingSize(this.getAccount);
+    const costOfSpendingChange =
+      changeSpendingSize * CoinSelector.LONG_TERM_FEERATE;
+
+    // cost of change = fee paid to create change +
+    // fee paid to spend that change in future
+    const costOfChange = this.getFee(changeOutputSize) + costOfSpendingChange;
+
+    let target = this.outputValue;
+    let size = await this.tx.estimateSize(this.getAccount);
+    const fee = this.getFee(size);
+    target += fee;
+
+    let bestSelection = [];
+    let bestWaste = consensus.MAX_MONEY;
+    let isChange = false; // true if bestSelection is BnB
+
+    // try to find a changeless solution using BnB
+    let result = this.selectBnB(target, costOfChange);
+    if (result && result.length > 0) {
+      bestSelection = result;
+      bestWaste = this.getWaste(result, costOfChange, target);
+      isChange = true;
```

Comment:
Should be `false` here, right? Also lets call this flag `hasChange`.

bnb: false
LL: true
SRD: true

...right?